### ISSUE TITLE
Serialization omits required properties with null values

### ISSUE THREAD

**Issue by lgolding** [microsoft/sarif-sdk#1740]

**Comment by eddynaka** [comment]
@lgolding , I would like to know if this is still a problem.

**Comment by lgolding** [comment]
@eddynaka, thank you for your contributions! I just ran the program above on the latest SARIF SDK (2.2.5), and yes, it is still a problem.

**Comment by eddynaka** [comment]
Hi @lgolding , i have some questions:
- in your example, when we do the first `DeserializeObject`, the `FullDescription` should be available with the `Text`? Right now, it's a null object.
- in your example, when we do the first `SerializeObject`, the `FullDescription` should be available with the `Text`? Right now, it's just showing as empty.

**Comment by lgolding** [comment]
Here is the "minimal repro file" from the issue description at the top:
```
{
  "$schema": "http://json.schemastore.org/sarif-2.1.0-rtm.4",
  "version": "2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "Bug1740",
          "rules": [
            {
              "id": "TST0001",
              "name": "Test rule",
              "fullDescription": {
                "text": null
              }
            }
          ]
        }
      },
      "results": [],
      "columnKind": "utf16CodeUnits"
    }
  ]
}
```
You see that `fullDescription.text` is present, but it is `null`. So when you do the first deserialize operation, the `FullDescription.Text` property is `null`, as you see.

Now, when you re-serialize, it becomes this:
```
{
  "$schema": "http://json.schemastore.org/sarif-2.1.0-rtm.4",
  "version": "2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "Bug1740",
          "rules": [
            {
              "id": "TST0001",
              "name": "Test rule",
              "fullDescription": {
              }
            }
          ]
        }
      },
      "results": [],
      "columnKind": "utf16CodeUnits"
    }
  ]
}
```

You see that now, `fullDescription.text` is completely absent, even though the schema says it is a required property. That is the bug.

Finally, when you do the second deserialization operation, it fails, because the required property `fullDescription.text` is absent from the JSON.

Does that help?

**Comment by eddynaka** [comment]
@lgolding , using your sample, I could achieve what you are saying doing this:
For the following code, I had to remove/always return true;
https://github.com/microsoft/sarif-sdk/blob/bb1ee25bda0d070691570c8a681827ae23d22097/src/Sarif/Core/MultiformatMessageString.cs#L6-L11

Doing that, we won't fail anymore. What do you think?

**Comment by lgolding** [comment]
I agree that `ShouldSerializeText` should always return `true`.

Do not change `ShouldSerializeMarkdown`, because the `markdown` property is optional. If it is empty or `null`, we do _not_ want to serialize it.

**Comment by michaelcfanning** [comment]
@lgolding, why are you suggesting this specific fix? We don't buy anything useful by persisting invalid SARIF. I think the real fix is to not emit 'fullDescription' itself if there's no useful interior data.

**Comment by eddynaka** [comment]
@michaelcfanning the problem here is that, if we don't change, if we have the fulldescription as null, you won't be able to serialize and desserialize. It will throw an exception. The change would solve that.

**Comment by michaelcfanning** [comment]
Yes. You should add a ShouldSerializeFullDescription() helper to Core\ReportingDescriptor.cs. Make sense? If you look at that code, you will see other examples that use special helpers like 'HasAtLeastOneNonNullValue'. We might require a new helper that also accounts for non-null but empty strings.

**Comment by eddynaka** [comment]
@lgolding / @michaelcfanning , doing as @michaelcfanning suggested, we can maintain that condition and create the new method in `Core\ReportingDescriptor.cs`, with that, the same test passes. Waiting for your comments before pushing another commit

**Comment by lgolding** [comment]
@eddynaka Sorry for delay in getting back to you. Yes, please do as @michaelcfanning suggests.

**Comment by eddynaka** [comment]
Hi @lgolding , it's already done :)

**Comment by lgolding** [comment]
Here's my opinion after all the thought we've put into this:

The problem occurs because when the deserializer sees
```
"fullDescription": {
    "text": null
}
```
... it says to itself "Great! This is valid. The `text` property is there.". But it's not really valid. The required value of the `text` property is `string` -- and `null` isn't a `string` in JSON; they're distinct data types (see the [JSON spec](https://www.json.org/json-en.html)).

So when we serialize, and `text` disappears altogether:
```
"fullDescription": {}
```
... we're really no worse off than before. The SARIF was invalid before, and it's still invalid. The bug is _not_ that we fail to deserialize the empty object; it's that we _succeeded_ in deserializing the object with a `null` value for `text`.

And that bug is in Newtonsoft. It sees the declaration of the `Text` property in the `MultiformatMessageString` class as a .NET `string`, and so it accepts a `null` value because in .NET `null` can stand in for a string. But it shouldn't.

I suggest we "won't fix" #1740, and if we're really ambitious, file a bug on Newtonsoft.

**Comment by lgolding** [comment]
... or maybe we introduce a "string converter" that intercepts deserialization of all string properties, and if the property is marked `Required` but it's `null`, we throw.
